packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js

```diff
@@ -0,0 +1,403 @@
+      act(click);
+      expect(calledCounter).toBe(5);
+      expect(button.innerHTML).toBe('5');
+
+      document.body.removeChild(container);
+    });
+
+    it('should flush effects recursively', () => {
+      function App() {
+        let [ctr, setCtr] = React.useState(0);
+        React.useEffect(() => {
+          if (ctr < 5) {
+            setCtr(x => x + 1);
+          }
+        });
+        return ctr;
+      }
+
+      const container = document.createElement('div');
+      act(() => {
+        ReactDOM.render(<App />, container);
+      });
+
+      expect(container.innerHTML).toBe('5');
+    });
+
+    it('detects setState being called outside of act(...)', () => {
+      let setValue = null;
+      function App() {
+        let [value, _setValue] = React.useState(0);
+        setValue = _setValue;
+        return (
+          <button id="button" onClick={() => setValue(2)}>
+            {value}
+          </button>
+        );
+      }
+      const container = document.createElement('div');
+      document.body.appendChild(container);
+      let button;
+      act(() => {
+        ReactDOM.render(<App />, container);
+        button = container.querySelector('#button');
+        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
+      });
+      expect(button.innerHTML).toBe('2');
+      expect(() => setValue(1)).toWarnDev([
+        'An update to App inside a test was not wrapped in act(...).',
+      ]);
+      document.body.removeChild(container);
+    });
+    describe('fake timers', () => {
+      beforeEach(() => {
+        jest.useFakeTimers();
+      });
+      afterEach(() => {
+        jest.useRealTimers();
+      });
+      it('lets a ticker update', () => {
+        function App() {
+          let [toggle, setToggle] = React.useState(0);
+          React.useEffect(() => {
+            let timeout = setTimeout(() => {
+              setToggle(1);
+            }, 200);
+            return () => clearTimeout(timeout);
+          }, []);
+          return toggle;
+        }
+        const container = document.createElement('div');
+
+        act(() => {
+          ReactDOM.render(<App />, container);
+        });
+        act(() => {
+          jest.runAllTimers();
+        });
+
+        expect(container.innerHTML).toBe('1');
+      });
+      it('can use the async version to catch microtasks', async () => {
+        function App() {
+          let [toggle, setToggle] = React.useState(0);
+          React.useEffect(() => {
+            // just like the previous test, except we
+            // use a promise and schedule the update
+            // after it resolves
+            sleep(200).then(() => setToggle(1));
+          }, []);
+          return toggle;
+        }
+        const container = document.createElement('div');
+
+        act(() => {
+          ReactDOM.render(<App />, container);
+        });
+        await act(async () => {
+          jest.runAllTimers();
```

Comment:
As a follow-up item, we should talk to Jest about a version of fake timers that encourages/requires you to wait for a new task. I'm guessing they haven't thought much about this problem because Facebook polyfills promises and transpiles async/await.

Reply:
I spoke to a couple of folks around here, and they mentioned they're looking at using https://nodejs.org/api/async_hooks.html